### PR TITLE
[Feat] (ABC-623) New Card component

### PR DESCRIPTION
**Pull request by lgilmet**
Doc: https://avonnicreator.atlassian.net/wiki/spaces/ABC/pages/2151186433/Card

### CODE REVIEW

File: src/modules/base/card/card.css [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+    flex: 1 1 auto;
+}
+.avonni-card__media-container {
+    flex: 1 4 auto;
+}
+.avonni-card__media-left {
+    flex-direction: row;
+}
+.avonni-card__media-right {
+    flex-direction: row-reverse;
```
Comment (nawel-aa) [inline src/modules/base/card/card.css:74]: [`slds-grid_reverse`](https://www.lightningdesignsystem.com/utilities/grid/#Right-to-Left)

File: src/modules/base/card/card.css [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+}
+.avonni-card__media-bottom,
+.avonni-card__media-center {
+    flex-direction: column-reverse;
```
Comment (nawel-aa) [inline src/modules/base/card/card.css:78]: [`slds-grid_vertical-reverse`](https://www.lightningdesignsystem.com/utilities/grid/#Bottom-to-Top)

File: src/modules/base/card/card.css [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+}
+.avonni-card__media-right .avonni-card__media-container {
+    position: relative;
+    align-self: auto;
+    height: auto;
+    object-fit: cover;
+    min-width: var(--avonni-card-right-media-width);
+    max-width: var(--avonni-card-right-media-width);
+}
+.avonni-card__media-background .avonni-card__media-container,
```
Comment (nawel-aa) [inline src/modules/base/card/card.css:187]: Everything except the `media-left`? If so, you can use `.avonni-card__body-container:not(.avonni-card__media-left) .avonni-card__media-container`.

File: src/modules/base/card/card.js [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+        'background',
+        'overlay'
+    ],
+    default: 'top'
+};
+
+export default class Card extends LightningElement {
+    _title;
+    _iconName;
+    _mediaPosition = 'top';
```
Comment (nawel-aa) [inline src/modules/base/card/card.js:53]: It should be set to `MEDIA_POSITIONS.default`

**Review by nawel-aa — CHANGES_REQUESTED** [review]
It works well. 👍 
On top of the other comments, maybe you could add a div with a max-width to the storybook wrapper and the examples? I'm afraid otherwise the examples will be too big in our doc website.

File: src/modules/base/card/__docs__/card.jsdoc.js [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+ * @name default
+ */
+
+/**
+ * @namespace stylingHooks
+ */
+/**
+ * @memberof stylingHooks
+ * @name --avonni-card-shadow
+ * @default '0 2px 2px 0 rgba(0, 0, 0, 0.1)'
```
[the remark this answers is not in the document]
Comment (lgilmet) [reply src/modules/base/card/__docs__/card.jsdoc.js:51]: Ok it just looked weird with the first string being highlighted and the rest looking like a comment :) but you are right.

File: src/modules/base/card/__docs__/card.jsdoc.js [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+ */
+/**
+ * @memberof stylingHooks
+ * @name --avonni-card-header-font-weight
+ * @default 700
+ * @type string
+ */
+/**
+ * @memberof stylingHooks
+ * @name --avonni-card-header-font-size
```
[the remark this answers is not in the document]
Comment (lgilmet) [reply src/modules/base/card/__docs__/card.jsdoc.js:85]: 👍

File: src/modules/base/card/__docs__/card.jsdoc.js [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+ */
+/**
+ * @memberof stylingHooks
+ * @name --avonni-card-shadow
+ * @default '0 2px 2px 0 rgba(0, 0, 0, 0.1)'
+ * @type string
+ */
+/**
+ * @memberof stylingHooks
+ * @name --avonni-card-text-color
```
[the remark this answers is not in the document]
Comment (lgilmet) [reply src/modules/base/card/__docs__/card.jsdoc.js:56]: 👍

File: src/modules/base/card/__docs__/card.jsdoc.js [hunk 8]
```diff
@@ -0,0 +1,236 @@
+/**
+ * @namespace slots
```
[the remark this answers is not in the document]
Comment (lgilmet) [reply src/modules/base/card/__docs__/card.jsdoc.js:2]: 👍

File: src/modules/base/card/card.js [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+     */
+
+    /**
+     * The title in the header of the card, right of the icon. The title attribute supersedes the title slot.
+     *
+     * @type {string}
+     * @public
+     */
+    @api
+    get title() {
```
[the remark this answers is not in the document]
Comment (lgilmet) [reply src/modules/base/card/card.js:96]: 🤦‍♂️

File: src/modules/base/card/card.js [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+    }
+
+    /**
+     * Source for the image or media.
+     *
+     * @type {string}
+     * @public
+     */
+    @api
+    get mediaSrc() {
```
[the remark this answers is not in the document]
Comment (lgilmet) [reply src/modules/base/card/card.js:127]: 🤦‍♂️

File: src/modules/base/card/card.js [hunk 11]
```diff
@@ -0,0 +1,397 @@
...
+        return this._mediaSrc;
+    }
+
+    set mediaSrc(value) {
+        this._mediaSrc = normalizeString(value);
+    }
+
+    /**
+     * Position of the media relative to the card.
+     * Valid values are "top", "center", "left", "right", "bottom", "background", and "overlay". Defaults to "top".
```
[the remark this answers is not in the document]
Comment (lgilmet) [reply src/modules/base/card/card.js:137]: 👍

File: src/modules/base/card/__docs__/card.stories.js [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+        mediaSrc: {
+            control: {
+                type: 'text'
+            },
+            description: 'Media link of the card image.',
+            table: {
+                type: { summary: 'string' }
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (lgilmet) [reply src/modules/base/card/__docs__/card.stories.js:85]: 👍

File: src/modules/base/card/__examples__/mediaBackground/mediaBackground.html [hunk 13]
```diff
@@ -0,0 +1,23 @@
+<template>
+    <avonni-card
+        title="Card title"
+        icon-name="custom:custom102"
+        media-src="https://react.lightningdesignsystem.com/assets/images/carousel/carousel-01.jpg"
+        media-position="top"
```
[the remark this answers is not in the document]
Comment (lgilmet) [reply src/modules/base/card/__examples__/mediaBackground/mediaBackground.html:6]: You understood the example name correctly.

File: src/modules/base/card/card.css [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/* outer card */
+
+article.avonni-card {
+    overflow: hidden;
+    display: flex;
```
[the remark this answers is not in the document]
Comment (lgilmet) [reply src/modules/base/card/card.css:37]: I can use slds-grid and slds-scrollable_none!

File: src/modules/base/card/card.css (hunk 3 above)
Comment (lgilmet) [reply src/modules/base/card/card.css:187]: Cool!

File: src/modules/base/card/__examples__/mediaLeft/mediaLeft.html [hunk 15]
```diff
@@ -0,0 +1,23 @@
+<template>
+    <avonni-card
+        title="Card title"
+        icon-name="custom:custom102"
+        media-src="https://react.lightningdesignsystem.com/assets/images/carousel/carousel-01.jpg"
+        media-position="left"
+        style="--avonni-card-left-media-width: 140px"
```
[the remark this answers is not in the document]
Comment (lgilmet) [reply src/modules/base/card/__examples__/mediaLeft/mediaLeft.html:7]: Using a smaller image solves this issue.

File: src/modules/base/card/card.css [hunk 16]
```diff
@@ -0,0 +1,257 @@
...
+
+    border-width: var(--avonni-card-sizing-border, 1px);
+    border-style: solid;
+    border-color: var(--avonni-card-color-border, #c9c9c9);
+    border-radius: var(--avonni-card-radius-border, 0.25rem);
+
+    box-shadow: var(--avonni-card-shadow, 0 2px 2px 0 rgba(0, 0, 0, 0.1));
+}
+
+.avonni-card__body-container {
```
[the remark this answers is not in the document]
Comment (lgilmet) [reply src/modules/base/card/card.css:56]: 👍

File: src/modules/base/card/card.css [hunk 17]
```diff
@@ -0,0 +1,257 @@
...
+    display: flex;
+    position: relative;
+    flex-direction: column;
+    flex: 1;
+    overflow: hidden;
+}
+
+.avonni-card__content-container {
+    flex: 1 1 auto;
```
[the remark this answers is not in the document]
Comment (lgilmet) [reply src/modules/base/card/card.css:65]: On image left and right the slds-col (flex: 1 1 auto) is overridden by slds-has-flexi-truncate (flex: 1 1 0%). But it is overridden by our own class.

File: src/modules/base/card/card.js [hunk 18]
```diff
@@ -0,0 +1,397 @@
...
+    showMedia = false;
+    showMediaSlot = false;
+    showMediaActionsSlot = false;
+    showTitleSlot = true;
+    showActionsSlot = false;
+    showDefaultSlot = true;
+    showFooterSlot = false;
+
+    renderedCallback() {
+        this.showMedia = !!this.mediaSrc || this.showMediaSlot;
```
[the remark this answers is not in the document]
Comment (lgilmet) [reply src/modules/base/card/card.js:65]: Fixed

File: src/modules/base/card/card.css (hunk 2 above)
Comment (lgilmet) [reply src/modules/base/card/card.css:78]: 👍

File: src/modules/base/card/card.css [hunk 19]
```diff
@@ -0,0 +1,257 @@
...
+    margin-bottom: var(--avonni-card-header-spacing-block-end, 0.75rem);
+    padding-left: var(--avonni-card-header-spacing-inline-start, 1rem);
+    padding-right: var(--avonni-card-header-spacing-inline-end, 1rem);
+    display: flex;
+    justify-content: space-between;
+    align-items: center;
+}
+
+.avonni-card__header-icon {
+    margin-right: 0.75rem;
```
[the remark this answers is not in the document]
Comment (lgilmet) [reply src/modules/base/card/card.css:221]: 👍

File: src/modules/base/card/card.css [hunk 20]
```diff
@@ -0,0 +1,257 @@
...
+}
+
+.avonni-card__header-title-section {
+    display: flex;
```
[the remark this answers is not in the document]
Comment (lgilmet) [reply src/modules/base/card/card.css:225]: 👍

File: src/modules/base/card/card.css (hunk 1 above)
Comment (lgilmet) [reply src/modules/base/card/card.css:74]: The slds-grid_reverse does not respect the flex shrink and grow.

File: src/modules/base/calendar/calendar.js [hunk 21]
```diff
@@ -1143,62 +1143,84 @@ export default class Calendar extends LightningElement {
 
         if (event.altKey) {
             if (event.keyCode === keyCodes.pageup) {
+                event.preventDefault();
+                event.stopPropagation();
```
Comment (nawel-aa) [inline src/modules/base/calendar/calendar.js:1147]: As discussed, you can remove all the `event.stopPropagation()`.

File: src/modules/base/card/__tests__/card.test.js [hunk 22]
```diff
@@ -41,13 +41,19 @@ describe('Card', () => {
         while (document.body.firstChild) {
             document.body.removeChild(document.body.firstChild);
         }
+        window.requestAnimationFrame.mockRestore();
```
Comment (nawel-aa) [inline src/modules/base/card/__tests__/card.test.js:44]: Mocking the `requestAnimationFrame` and using fake timers is not really a problem, but it's only necessary if you actually use `requestAnimationFrame` or timeouts. 🙂

File: src/modules/base/card/card.js [hunk 23]
```diff
@@ -211,7 +196,7 @@ export default class Card extends LightningElement {
      */
     get defaultSlot() {
         return this.template.querySelector(
-            'slot[data-element-id="avonni-card-center-default-slot"]'
+            'slot[data-element-id="avonni-card-default-slot"], slot[data-element-id="avonni-card-center-default-slot"]'
```
Comment (nawel-aa) [inline src/modules/base/card/card.js:199]: Ha! I didn't know you could chain the selectors like that! 😃

File: src/modules/base/card/card.js [hunk 24]
```diff
@@ -258,20 +237,22 @@ export default class Card extends LightningElement {
...
             this.title ||
-            !!this.iconName ||
+            this.iconName ||
             this.showActionsSlot
         );
     }
 
     /**
-     * Card body classes
+     * Card body classes ... slds-grid_vertical
```
Comment (nawel-aa) [inline src/modules/base/card/card.js:246]: I'm assuming this is a copy-paste mistake, and not a very suspenseful and mysterious description. 😄

**Review by nawel-aa — CHANGES_REQUESTED** [review]

File: src/modules/base/card/card.js [hunk 25]
```diff
@@ -0,0 +1,332 @@
...
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import { LightningElement, api } from 'lwc';
+import { classSet } from 'c/utils';
+import { normalizeString } from '../utilsPrivate/normalize';
```
[the remark this answers is not in the document]
Comment (lgilmet) [reply src/modules/base/card/card.js:35]: 👍

File: src/modules/base/card/card.js (hunk 24 above)
Comment (lgilmet) [reply src/modules/base/card/card.js:246]: 🤦‍♂️

File: src/modules/base/card/card.js (hunk 23 above)
Comment (lgilmet) [reply src/modules/base/card/card.js:199]: This one is useful :P

File: src/modules/base/card/card.js [hunk 26]
```diff
@@ -0,0 +1,332 @@
...
+        this.showActionsSlot =
+            this.actionsSlot &&
+            this.actionsSlot.assignedElements().length !== 0;
+        this.showDefaultSlot =
+            (this.defaultSlot &&
+                this.defaultSlot.assignedElements().length !== 0) ||
+            (this.defaultSlot &&
+                this.defaultSlot.innerText &&
+                this.defaultSlot.innerText.trim().length !== 0);
+        this.showTitleSlot =
```
[the remark this answers is not in the document]
Comment (lgilmet) [reply src/modules/base/card/card.js:108]: Of course thanks

File: src/modules/base/card/__docs__/card.stories.js [hunk 27]
```diff
@@ -82,23 +81,24 @@ export default {
...
 };
 
 const Template = (args) => Card(args);
 
 export const Base = Template.bind({});
 Base.args = {
-    title: 'Card Title',
-    iconName: 'custom:custom102',
-    mediaPosition: 'left'
+    // title: 'Card Title',
```
[the remark this answers is not in the document]
Comment (lgilmet) [reply src/modules/base/card/__docs__/card.stories.js:94]: I gotta start doing that, reviewing the changes in isolation. 🤦‍♂️

File: src/modules/base/card/card.html [hunk 28]
```diff
@@ -0,0 +1,121 @@
...
+                        <h2
+                            if:true={title}
+                            title={title}
+                            class="avonni-card__title-text slds-truncate"
+                            data-element-id="avonni-card-title"
+                        >
+                            {title}
+                        </h2>
+
+                        <template if:true={showTitleSlot}>
```
[the remark this answers is not in the document]
Comment (lgilmet) [reply src/modules/base/card/card.html:95]: Good catch

File: src/modules/base/card/__docs__/card.stories.js [hunk 29]
```diff
@@ -60,8 +60,7 @@ export default {
             ],
             description: 'Media position.',
             table: {
-                type: { summary: 'string' },
-                defaultValue: { summary: 'top' }
```
[the remark this answers is not in the document]
Comment (lgilmet) [reply src/modules/base/card/__docs__/card.stories.js]: Ah I see it now. I wondered where that "table" was.

File: src/modules/base/card/__tests__/card.test.js (hunk 22 above)
Comment (lgilmet) [reply src/modules/base/card/__tests__/card.test.js:44]: I tried solving a problem and forgot it there.

File: src/modules/base/card/card.css [hunk 30]
```diff
@@ -0,0 +1,257 @@
...
+}
+
+/* card header */
+
+.avonni-card__header {
+    padding-top: var(--avonni-card-header-spacing-block-start, 0.75rem);
+    margin-bottom: var(--avonni-card-header-spacing-block-end, 0.75rem);
+    padding-left: var(--avonni-card-header-spacing-inline-start, 1rem);
+    padding-right: var(--avonni-card-header-spacing-inline-end, 1rem);
+    display: flex;
```
[the remark this answers is not in the document]
Comment (lgilmet) [reply src/modules/base/card/card.css:199]: 👍

File: src/modules/base/calendar/calendar.js (hunk 21 above)
Comment (lgilmet) [reply src/modules/base/calendar/calendar.js:1147]: 👍

**Review by nawel-aa — APPROVED** [review]
🎉

**Pull request merged**